src/percy-agent-client/dom.ts

```diff
@@ -0,0 +1,210 @@
+   * form controls to make sure they persist in snapshots.
+   *
+   */
+  private serializeInputElements(clonedDOM: HTMLDocument): HTMLDocument {
+    const formNodes = this.originalDOM.querySelectorAll('input, textarea')
+    const formElements = Array.from(formNodes)
+
+    formElements.forEach((elem: any) => {
+      const inputId = elem.attributes['data-percy-element-id'].value
+      const selector = `[data-percy-element-id="${inputId}"]`
+      const cloneEl = clonedDOM.querySelector(selector) as HTMLInputElement
+
+      switch (elem.type) {
+        case 'checkbox':
+        case 'radio':
+          if (elem.checked && !elem.hasAttribute('checked')) {
+            cloneEl!.setAttribute('checked', '')
+          }
+          break
+        case 'textarea':
+          // setting text or value does not work but innerText does
+          if (elem.innerText !== elem.value) {
+            cloneEl!.innerText = elem.value
+          }
+        default:
+          if (!elem.getAttribute('value')) {
+            cloneEl!.setAttribute('value', elem.value)
+          }
+      }
+    })
+
+    return clonedDOM
+  }
+
+  /**
+   * Capture in-memory styles & serialize those styles into the cloned DOM.
+   *
+   * Without this, applications that use packages like `styled-components` will be
+   * missing styles and appear broken. The CSSOM provides a way to create CSS
+   * that only lives in memory (not an asset or in the DOM).
+   *
+   */
+  private serializeCssOm(documentClone: HTMLDocument) {
+    const styleSheets = Array.from(this.originalDOM.styleSheets)
+
+    styleSheets.forEach((styleSheet: any) => {
+      // Make sure it has a rulesheet, does NOT have a href (no external stylesheets),
+      // and isn't already in the DOM.
+      function isCSSOM() {
+        const hasHref = styleSheet.href
+        const ownerNode = styleSheet.ownerNode as HTMLElement
+        const hasStyleInDom =
+          ownerNode.innerText && ownerNode.innerText.length > 0
+
+        return !hasHref && !hasStyleInDom && styleSheet.cssRules
+      }
+
+      if (isCSSOM()) {
+        const $style = documentClone.createElement('style')
+        const cssRules = Array.from(styleSheet.cssRules)
+        const serializedStyles = cssRules.reduce((prev: string, cssRule: any) => {
+          return prev + cssRule.cssText
+        }, '')
+
+        // Append the serialized styles to the cloned document
+        $style.type = 'text/css'
+        $style.setAttribute('data-percy-cssom-serialized', 'true')
+        $style.innerHTML = serializedStyles
+        // TODO, it'd be better if we appended it right after the ownerNode in the clone
+        documentClone.head.appendChild($style)
+      }
+    })
+
+    return documentClone
+  }
+
+  /**
+   * A single place to mutate the origial DOM. This should be the last resort!
+   * This will change the customers DOM and have a possible impact on the
+   * customers application.
+   *
+   */
```

Comment:
May also want to include comments above saying currently we're only using it to add data-percy-element-id's to inputs which should be safe.  Also that this mutation can apply to the same original DOM multiple times within one test/snapshot set.  This can be dangerous, which is why we suggest strictly minimizing changes made here.